specs/GetSecurityDescriptorForPackageFamilyNames/GetSecurityDescriptorForPackageFamilyNames.md

```diff
@@ -0,0 +1,49 @@
+# Easing and streamlining named object sharing between packaged and Win32
+
+This feature is a flat C API that takes a list of Package Family Names (PFNs) and access masks, and returns a security descriptor
+
+# Background
+
+Named objects (mutexes, events, waitable timers) must be ACLed during creation in order to be shared with packaged applications.
+
+Currently, [our guidance](https://docs.microsoft.com/en-us/windows/uwp/communication/sharing-named-objects) for this scenario is to write a large amount of complex, cumbersome, and largely boilerplate code. ([See here for a taste](https://docs.microsoft.com/en-us/windows/win32/api/securityappcontainer/nf-securityappcontainer-getappcontainernamedobjectpath#examples)).
+
+If we are to meet developers where they are, we need to reduce the friction of Win32 and UWP interop, and make it as easy as possible to share named objects.
+
+# Description
+
+Add a flat C API that takes an array of PFNs and access masks, and produces the needed `SECURITY_ATTRIBUTES` object.
+
+# Examples
+
+```c++
+#include <wil/resource.h>
+
+wil::unique_event CreateShareableEvent(PCWSTR name)
+{
+    wil::unique_hlocal_ptr pSD;
+    PCWSTR packageNames[1] = { L"Contoso.Test.App_12345678" };
+    DWORD accessMasks[1] = { ACCESS_MODIFY_STATE };
+
+    THROW_IF_FAILED(GetSecurityDescriptorForPackageFamilyNames(
+        1, packageNames, accessMasks, &pSD);
+
+    wil::unique_event result;
+    result.create(wil::EventOptions::None, name, pSD.get());
+    return result;
+}
+```
+
+# API Details
+
+```c
+STDAPI GetSecurityDescriptorForPackageFamilyNames(
+    DWORD                 cCountOfPackageFamilyNames,
+    _In_Reads_(cCountOfPackageFamilyNames)
+      const PCWSTR*         pListOfPackageFamilyNames,
+    _In_Reads_(cCountOfPackageFamilyNames)
```

Comment:
So if I want to grant 10 apps the same level of permissions, I have to create a list with 10 copies of the same permissions?

I think we should definitely support assigning the same permissions to multiple apps without having to list those permissions once for each app. Though also supporting a list of permissions for each app is nice as well to save the developer from having to write the complicated code in that scenario.

To support both, would that be two separate methods?

Reply:
I've added an "API Notes" section that discusses the current API proposal, as well as two alternatives. I chose this one because it carried the smallest downside, in my opinion.

The alternatives either sacrificed valuable functionality (only accept a single permission), introduced complexity to the API (add a single permission, and make the array of permissions optional), or introduced a doc/naming/developer mental model penalty (two nearly identical APIs, one takes a single permission, the other takes an array of permissions). I believe the current proposal sacrifices the least, but I'm willing to listen to counterarguments.

Reply:
I definitely don't like 2 separate methods if we can avoid it. Option 2 of having both arguments and using the list if it's provided doesn't seem that complicated to me, as long as that is a relatively standard design pattern for flat C APIs. C# would overload on type for this. Presumably this is the flat C equivalent of that?

Reply:
While experimenting with the implementation, it turns out this only works if we also grant access at the user level. This means we also need an access mask for the user in the custom case.

I also want to tweak this, because when we pass two separate arrays, the caller needs to ensure they are both the same length. A more robust design would pass an array of `struct { UINT32 accessMask; PCWSTR name };`.

Reply:
Could you clarify how this changes the API?

Reply:
The API description and usage example have already been updated. Could you help me understand what you want in addition to that?

Reply:
I see that now. We never resolved the discussion on one access mask provided/applied to multiple apps (and now also the user), and this design takes us further from that. Are we giving up on simplifying that scenario with the new design?

Reply:
Yeah, I've abandoned that idea. For the per-app access masks to be optional, the caller needs to pass two arrays (instead of a single array of name/mask pairs).
1. This provides opporunities for callers to get it wrong and pass different-sized arrays.
2. If the array size is determined at runtime, that also means two separate heap allocations.

To me, this outweighs the benefits of having a slightly simpler calling pattern; the sample code hardly suffers from omitting the "simple" flavor.

Reply:
Ok. The concern was always granting access to multiple apps, not just one, so the sample code for 1 app wouldn't really suffer in any case. Though I do agree that coupling an AppContainerName to an access mask with a struct has its advantages so am ok with this.